File: ValidParantheses.py
--------------------------Valid parantheses------------------------------------------
# Time Complexity : O(N) as N is length of s 
# Space Complexity : O(1) 
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
# we wie through ll use stack and dict, we will have parentheses in dict and iterate through s and check if the char in dict
# or not, if it is open parentheses, we will append to the stack, else we will check if the top of the stack is same paranthesis or not
#If not we will return false else in the last we will check if there are any elements in the stack, return False
# else return True 


class Solution:
    def isValid(self, s: str) -> bool:
        if not s:
            return True
        
        d = {'(':')', '{':'}', '[':']'}
        stack = []
        for i in s:
            if i in d:
                stack.append(d[i])
            elif not stack or stack.pop() != i:
                return False
        return len(stack) == 0

File: valid_parantheses.cpp
// Time Complexity : O(n)
// Space Complexity :  O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this - 

class Solution {
public:
    bool isValid(string s) {
        
        int len = s.length();
        int i;
        
        stack<char> str_st;
        
        for(i = 0; i < len; i++)
        {
            // Opening
            if(s[i] == '(' || s[i] == '{' || s[i] == '[')
            {
                str_st.push(s[i]);
            }
            // Closing
            else
            {
                // Non Empty stack
                if(!str_st.empty())
                {
                if(str_st.top() == '(' && s[i] == ')')
                    str_st.pop();
                else if(str_st.top() == '{' && s[i] == '}')
                    str_st.pop();
                else if(str_st.top() == '[' && s[i] == ']')
                    str_st.pop();
                // Incorrect matching of parantheses
                else
                    return false;
                }
                // EMpty stack
                else
                    return false;
            }
        }
        
        // True only if stack empty
        if(str_st.empty())
            return true;
        else
            return false;
    }
};